# Data cleaning and transformation  
This notebook deals with cleaning the dataset and transforming it for modelling.  
As usual we will start by importing libraries

In [1]:
import seaborn as sns
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


Reading in the Data for the Case Level information.

In [2]:
df=pd.read_csv('data/database.csv')
print(df.shape)

/Users/J/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(638454, 24)


Lets see what out dataset Looks like. We transposed it so we could see all the column names.

In [3]:
df.head(5).transpose()

0                       1  \
Record ID                                          1                       2   
Agency Code                                  AK00101                 AK00101   
Agency Name                                Anchorage               Anchorage   
Agency Type                         Municipal Police        Municipal Police   
City                                       Anchorage               Anchorage   
State                                         Alaska                  Alaska   
Year                                            1980                    1980   
Month                                        January                   March   
Incident                                           1                       1   
Crime Type                    Murder or Manslaughter  Murder or Manslaughter   
Crime Solved                                     Yes                     Yes   
Victim Sex                                      Male                    Male   
Victim Age                                        14                      43   
Victim Race            Native American/Alaska Native                   White   
Victim Ethnicity                             Unknown                 Unknown   
Perpetrator Sex                                 Male                    Male   
Perpetrator Age                                   15                      42   
Perpetrator Race       Native American/Alaska Native                   White   
Perpetrator Ethnicity                        Unknown                 Unknown   
Relationship                            Acquaintance            Acquaintance   
Weapon                                  Blunt Object           Strangulation   
Victim Count                                       0                       0   
Perpetrator Count                                  0                       0   
Record Source                                    FBI                     FBI   

                                                   2                       3  \
Record ID                                          3                       4   
Agency Code                                  AK00101                 AK00101   
Agency Name                                Anchorage               Anchorage   
Agency Type                         Municipal Police        Municipal Police   
City                                       Anchorage               Anchorage   
State                                         Alaska                  Alaska   
Year                                            1980                    1980   
Month                                          March                   April   
Incident                                           2                       1   
Crime Type                    Murder or Manslaughter  Murder or Manslaughter   
Crime Solved                                      No                     Yes   
Victim Sex                                    Female                    Male   
Victim Age                                        30                      43   
Victim Race            Native American/Alaska Native                   White   
Victim Ethnicity                             Unknown                 Unknown   
Perpetrator Sex                              Unknown                    Male   
Perpetrator Age                                    0                      42   
Perpetrator Race                             Unknown                   White   
Perpetrator Ethnicity                        Unknown                 Unknown   
Relationship                                 Unknown            Acquaintance   
Weapon                                       Unknown           Strangulation   
Victim Count                                       0                       0   
Perpetrator Count                                  0                       0   
Record Source                                    FBI                     FBI   

                                                 

Df.info shows us the datatypes of our columns.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 638454 entries, 0 to 638453
Data columns (total 24 columns):
Record ID                638454 non-null int64
Agency Code              638454 non-null object
Agency Name              638454 non-null object
Agency Type              638454 non-null object
City                     638454 non-null object
State                    638454 non-null object
Year                     638454 non-null int64
Month                    638454 non-null object
Incident                 638454 non-null int64
Crime Type               638454 non-null object
Crime Solved             638454 non-null object
Victim Sex               638454 non-null object
Victim Age               638454 non-null int64
Victim Race              638454 non-null object
Victim Ethnicity         638454 non-null object
Perpetrator Sex          638454 non-null object
Perpetrator Age          638454 non-null object
Perpetrator Race         638454 non-null object
Perpetrator Ethnicity    6384

### Changing columns to numeric
As we can see some of these columns can easily be changed to numeric columns  
Lets start with the month column and use a dictionary and mapping to turn it to numeric

In [5]:
month = {'01':'January',
'02':'February',
'03':'March',
'04':'April',
'05':'May',
'06':'June',
'07':'July',
'08':'August',
'09':'September',
'10':'October',
'11':'November',
'12':'December'}
#This dictionary needs to be inverted 
inv_map = {v: k for k, v in month.items()}
#mapping 
df['Month']=df['Month'].map(inv_map)
#changing string to integer
df['Month']=df['Month'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 638454 entries, 0 to 638453
Data columns (total 24 columns):
Record ID                638454 non-null int64
Agency Code              638454 non-null object
Agency Name              638454 non-null object
Agency Type              638454 non-null object
City                     638454 non-null object
State                    638454 non-null object
Year                     638454 non-null int64
Month                    638454 non-null int64
Incident                 638454 non-null int64
Crime Type               638454 non-null object
Crime Solved             638454 non-null object
Victim Sex               638454 non-null object
Victim Age               638454 non-null int64
Victim Race              638454 non-null object
Victim Ethnicity         638454 non-null object
Perpetrator Sex          638454 non-null object
Perpetrator Age          638454 non-null object
Perpetrator Race         638454 non-null object
Perpetrator Ethnicity    63845

We can see that the 'Perpetrator Age' column is object type. we will change it to integer.

In [6]:
df["Perpetrator Age"]=pd.to_numeric(df["Perpetrator Age"],errors='coerce')

In [7]:
d = {'No': 0, 'Yes': 1}
df['Crime Solved'] = df['Crime Solved'].map(d).fillna(df['Crime Solved'])

In [8]:
df.describe()

Record ID           Year          Month       Incident  \
count  638454.00000  638454.000000  638454.000000  638454.000000   
mean   319227.50000    1995.801102       6.585622      22.967924   
std    184305.93872       9.927693       3.412824      92.149821   
min         1.00000    1980.000000       1.000000       0.000000   
25%    159614.25000    1987.000000       4.000000       1.000000   
50%    319227.50000    1995.000000       7.000000       2.000000   
75%    478840.75000    2004.000000       9.000000      10.000000   
max    638454.00000    2014.000000      12.000000     999.000000   

        Crime Solved     Victim Age  Perpetrator Age   Victim Count  \
count  638454.000000  638454.000000    638453.000000  638454.000000   
mean        0.701964      35.033512        20.322697       0.123334   
std         0.457396      41.628306        17.886842       0.537733   
min         0.000000       0.000000         0.000000       0.000000   
25%         0.000000      22.000000         0.000000       0.000000   
50%         1.000000      30.000000        21.000000       0.000000   
75%         1.000000      42.000000        31.000000       0.000000   
max         1.000000     998.000000        99.000000      10.000000   

       Perpetrator Count  
count      638454.000000  
mean            0.185224  
std             0.585496  
min             0.000000  
25%             0.000000  
50%             0.000000  
75%             0.000000  
max            10.000000

In [9]:
df.isna().sum()

Record ID                0
Agency Code              0
Agency Name              0
Agency Type              0
City                     0
State                    0
Year                     0
Month                    0
Incident                 0
Crime Type               0
Crime Solved             0
Victim Sex               0
Victim Age               0
Victim Race              0
Victim Ethnicity         0
Perpetrator Sex          0
Perpetrator Age          1
Perpetrator Race         0
Perpetrator Ethnicity    0
Relationship             0
Weapon                   0
Victim Count             0
Perpetrator Count        0
Record Source            0
dtype: int64

In [10]:
df[df['Perpetrator Age'].isna()]

Record ID Agency Code Agency Name       Agency Type   City     State  \
634666     634667     OK07205       Tulsa  Municipal Police  Tulsa  Oklahoma   

        Year  Month  Incident              Crime Type  ...  Victim Ethnicity  \
634666  2014      6       104  Murder or Manslaughter  ...      Not Hispanic   

       Perpetrator Sex  Perpetrator Age Perpetrator Race  \
634666         Unknown              NaN          Unknown   

       Perpetrator Ethnicity Relationship   Weapon Victim Count  \
634666               Unknown      Unknown  Handgun            0   

       Perpetrator Count Record Source  
634666                 0           FBI  

[1 rows x 24 columns]

In [11]:
df.dropna(inplace=True)

In [12]:
df.isna().sum()

Record ID                0
Agency Code              0
Agency Name              0
Agency Type              0
City                     0
State                    0
Year                     0
Month                    0
Incident                 0
Crime Type               0
Crime Solved             0
Victim Sex               0
Victim Age               0
Victim Race              0
Victim Ethnicity         0
Perpetrator Sex          0
Perpetrator Age          0
Perpetrator Race         0
Perpetrator Ethnicity    0
Relationship             0
Weapon                   0
Victim Count             0
Perpetrator Count        0
Record Source            0
dtype: int64

Now let clean up the Victim Age columns. A value of 998 represents victims whose age was not reported, usually because the victim was unidentified and the age was unknown. Lets see how many cases have unknown Victim age

In [13]:
df[df['Victim Age']>100]['Victim Age'].count()

974

As seen above its a small number compare to our dataset and we can just take it out and keep known values

In [14]:
df=df[df['Victim Age']<100]

### Coding agency and location information   
Our next concern is how to include the agency information into our data. There are tens of thousands of agencies and one-hot encoding would not make sense for them. We will include external data for that. The clearance dataset is from Uniform Crime Report data summarizing all homicides and homicide clearances reported from 1965 to the present.  
ORI is the Agency code 

In [15]:
clearance=pd.read_csv('data/UCR65_18.csv')
clearance.head()

ORI       Name  YEAR  MRD  CLR   State         County     Agency
0  AK00101  ANCHORAGE  1965    7    6  Alaska  Anchorage, AK  Anchorage
1  AK00101  ANCHORAGE  1966   18   16  Alaska  Anchorage, AK  Anchorage
2  AK00101  ANCHORAGE  1967    1    1  Alaska  Anchorage, AK  Anchorage
3  AK00101  ANCHORAGE  1968    7    5  Alaska  Anchorage, AK  Anchorage
4  AK00101  ANCHORAGE  1969    7    4  Alaska  Anchorage, AK  Anchorage

In [16]:
clearance.shape

(163087, 8)

Lets add a column that calculates the clearance rate for each agency for that particular year. We will need to need to drop the row where clearance rate is infinite values.

In [17]:
clearance["rate"]=clearance["CLR"]/clearance["MRD"]

In [18]:
indx=list(np.where(clearance.rate >= np.finfo(np.float64).max)[0])
clearance.drop(clearance.index[indx],inplace=True)

Lets merge the clearance dataset with out original one. We will perform a left join.

In [19]:
full_df = df.merge(clearance[['ORI','YEAR','rate']], how='left', left_on=['Agency Code', 'Year'], right_on=['ORI', 'YEAR'])


In [20]:
full_df.shape

(637479, 27)

In [21]:
full_df.isna().sum()

Record ID                   0
Agency Code                 0
Agency Name                 0
Agency Type                 0
City                        0
State                       0
Year                        0
Month                       0
Incident                    0
Crime Type                  0
Crime Solved                0
Victim Sex                  0
Victim Age                  0
Victim Race                 0
Victim Ethnicity            0
Perpetrator Sex             0
Perpetrator Age             0
Perpetrator Race            0
Perpetrator Ethnicity       0
Relationship                0
Weapon                      0
Victim Count                0
Perpetrator Count           0
Record Source               0
ORI                      6222
YEAR                     6222
rate                     6222
dtype: int64

As we performed a Left join , we have a few nulls where the data did not match. we will fill the null for clearance rate with the mean of the clearance rate for the respective state. 

In [22]:
full_df['rate'] = full_df['rate'].fillna(full_df.groupby('State')['rate'].transform('mean'))

We will drop the duplicate columns

In [23]:
full_df.drop(labels=['ORI','YEAR'],axis=1,inplace=True)

In [24]:
full_df.isna().sum()

Record ID                0
Agency Code              0
Agency Name              0
Agency Type              0
City                     0
State                    0
Year                     0
Month                    0
Incident                 0
Crime Type               0
Crime Solved             0
Victim Sex               0
Victim Age               0
Victim Race              0
Victim Ethnicity         0
Perpetrator Sex          0
Perpetrator Age          0
Perpetrator Race         0
Perpetrator Ethnicity    0
Relationship             0
Weapon                   0
Victim Count             0
Perpetrator Count        0
Record Source            0
rate                     0
dtype: int64

In [25]:
full_df.shape

(637479, 25)

We will write this dataframe to csv to use in our modelling notebook

In [26]:
full_df.to_csv('data/dataset_w_clearance.csv')